clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java

```diff
@@ -662,7 +664,7 @@ private void ensureOpenForRecordBatchWrite() {
      */
     private int estimatedBytesWritten() {
         if (compressionType == CompressionType.NONE) {
-            return buffer().position();
+            return buffer().position() - initialPosition;
```

Comment:
Hmm.. We'd need to add on the size of the batch header, which we should probably be doing for the compressed path anyway. Let me see how this looks.